sdk/src/types/block/output/feature/metadata.rs

```diff
@@ -1,103 +1,250 @@
+pub(crate) type MetadataFeatureEntryCount = BoundedU8<1, { u8::MAX }>;
+pub(crate) type MetadataFeatureKeyLength = BoundedU8<1, { u8::MAX }>;
+pub(crate) type MetadataFeatureValueLength = BoundedU16<0, { u16::MAX }>;
+
+type MetadataBTreeMapPrefix = BTreeMapPrefix<
+    BoxedSlicePrefix<u8, MetadataFeatureKeyLength>,
+    BoxedSlicePrefix<u8, MetadataFeatureValueLength>,
+    MetadataFeatureEntryCount,
+>;
+
+type MetadataBTreeMap =
+    BTreeMap<BoxedSlicePrefix<u8, MetadataFeatureKeyLength>, BoxedSlicePrefix<u8, MetadataFeatureValueLength>>;
 
 /// Defines metadata, arbitrary binary data, that will be stored in the output.
-#[derive(Clone, Eq, PartialEq, Ord, PartialOrd, Hash, packable::Packable)]
-#[packable(unpack_error = Error, with = |err| Error::InvalidMetadataFeatureLength(err.into_prefix_err().into()))]
-pub struct MetadataFeature(
-    // Binary data.
-    pub(crate) BoxedSlicePrefix<u8, MetadataFeatureLength>,
-);
+#[derive(Clone, Eq, PartialEq, Ord, PartialOrd, Hash)]
+pub struct MetadataFeature(MetadataBTreeMapPrefix);
+
+fn verify_packable<const VERIFY: bool>(feature: &MetadataFeature) -> Result<(), Error> {
+    verify_keys_packable::<VERIFY>(feature)?;
+    verify_byte_length_packable::<VERIFY>(feature.packed_len())?;
+    Ok(())
+}
+
+fn verify_keys_packable<const VERIFY: bool>(feature: &MetadataFeature) -> Result<(), Error> {
+    if VERIFY {
+        for key in feature.0.keys() {
+            if !key.iter().all(|c| c.is_ascii_graphic()) {
+                return Err(Error::NonGraphicAsciiMetadataKey(key.to_vec()));
+            }
+        }
+    }
+    Ok(())
+}
+
+fn verify_byte_length_packable<const VERIFY: bool>(len: usize) -> Result<(), Error> {
+    if VERIFY
+        && !MetadataFeature::BYTE_LENGTH_RANGE
+            .contains(&u16::try_from(len).map_err(|e| Error::InvalidMetadataFeature(e.to_string()))?)
+    {
+        return Err(Error::InvalidMetadataFeature(format!(
+            "Out of bounds byte length: {len}"
+        )));
+    }
+    Ok(())
+}
 
 impl MetadataFeature {
     /// The [`Feature`](crate::types::block::output::Feature) kind of [`MetadataFeature`].
     pub const KIND: u8 = 2;
-    /// Valid lengths for a [`MetadataFeature`].
-    pub const LENGTH_RANGE: RangeInclusive<u16> = 1..=8192;
+    /// Valid byte lengths for a [`MetadataFeature`].
+    pub const BYTE_LENGTH_RANGE: RangeInclusive<u16> = 1..=8192;
 
     /// Creates a new [`MetadataFeature`].
     #[inline(always)]
-    pub fn new(data: impl Into<Vec<u8>>) -> Result<Self, Error> {
-        Self::try_from(data.into())
+    pub fn new(data: impl IntoIterator<Item = (String, Vec<u8>)>) -> Result<Self, Error> {
+        let mut builder = Self::build();
+        for (k, v) in data {
+            builder.insert(k, v);
+        }
+        builder.finish()
+    }
+
+    pub fn build() -> MetadataFeatureMap {
+        Default::default()
+    }
+
+    pub fn to_map(&self) -> MetadataFeatureMap {
```

Comment:
should `build()` and `to_map` be documented?

Reply:
Done